### DISCUSSION TITLE
Extract asset from .assets archive

### DISCUSSION THREAD

**Discussion by martanius** [nesrak1/AssetsTools.NET#134] [Q&A] [answered]
Hello.

I need to replace a specific file in an .asset archive but before I do that I need to verify the file that is already in the archive. So I'd need to get the actual file or atleast byte[] array of it.

I know about the BundleHelper.LoadAssetDataFromBundle but that only works for bundle files. I need the exact same thing but for asset files and there is no such method in AssetHelper class.

**Comment by nesrak1** [comment]
> replace a specific file in an .asset archive 

Do you mean an .assets file? Are you asking how to check if an asset exists in the assets file? That depends on how you want to identify it. Do you want to find it by path id? Do you want to find it by name? If you try to find it by name, you may run into collisions since names don't have to be unique.

**Reply by martanius** [reply]
Thank you for reply.

Yeah, sorry, I mean .assets file. The identification of the asset inside the assets file is not a problem.

But I need to compare the content of the asset with the one I'd be replacing it with.
So **I need to know if the asset I want to import into the archive (or rather replace) is different from the one that is already in the archive.**

I could compare the size but to do more precise comparison, it would be better if I could get the byte[] of the asset. But I see no way of getting that. Only the BundleHelper offers such method, but this is .assets archive.

In other words, how to export an asset that is inside .assets archive?

**Reply by nesrak1** [reply, marked as answer]
I don't think there's a built in method to export an asset to byte array, but you can use this code from UABEA: https://github.com/nesrak1/UABEANext/blob/fcb4f8a0313e341121bb98e6bc4bd3b4d7417afa/UABEANext4/Logic/ImportExport/AssetExport.cs#L20-L33 and make `_stream` a new MemoryStream or something.

`DumpRawAsset(afile.Reader, info.GetAbsoluteByteOffset(afile), info.ByteSize);`

**Reply by nesrak1** [reply]
You can also cheat and do
```
var baseField = manager.GetBaseField(afileInst, inf);
var bytes = baseField.WriteToByteArray();
```

**Reply by martanius** [reply]
Thank you. I'll do that :)

**Reply by martanius** [reply]
> I don't think there's a built in method to export an asset to byte array, but you can use this code from UABEA: https://github.com/nesrak1/UABEANext/blob/fcb4f8a0313e341121bb98e6bc4bd3b4d7417afa/UABEANext4/Logic/ImportExport/AssetExport.cs#L20-L33 and make `_stream` a new MemoryStream or something.
> 
> `DumpRawAsset(afile.Reader, info.GetAbsoluteByteOffset(afile), info.ByteSize);`

Actually this one did the trick.

The "var bytes = baseField.WriteToByteArray();" only returned array of 4 bytes or something.
Once again, thank you :)
